[noahang] iP

noahang opened the pull request:
### SirDuke

> "There are two kinds of worries – those you can do something about and those you can't. Don't spend any time on the latter" – Duke Ellington [(source)](https://www.blinkist.com/magazine/posts/timeless-wisdom-duke-ellington-top-10-quotes)

Modeled after the speech of jazz legend  Duke Ellington 🎹, SirDuke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- _SUPER FAST_ to use


All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 
5. And it is FREE!

Features:

- [x]  Managing tasks
- [x]  Managing deadlines (coming soon)
- [x]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method found in `SirDuke.java`:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

noahang committed:
Rename Duke class files to SirDuke, add greet and exit functions.

Renamed Duke in README to  SirDuke.

Renamed and personalised System.out.println that prints greeting message.

Add SirDuke.sayBye() method, that is called when program exits.

noahang committed:
Create SirDuke.start() method, add Echo functionality, fix style, remove logo

-SirDuke.start() method created to contain welcome message printing and user input reading from console

-SirDuke.start() method also contains Echo functionality, and calls SirDuke.sayBye() when "bye" command is received

-horizontalLine variable is a constant, so it's variable name has been changed to HORIZONTAL_LINE
-print messages have been updated to follow style guide for split strings
-logo variable has been removed as it is no longer necessary

noahang committed:
Add Task.java, ToDoList.java, Format.java
SirDuke.java: Move printing of commands to ToDoList class, add switch statement for command execution

Each task is as string.

Strings cannot encapsulate whether the task was completed or not.

Let's create a Task class, Task.java to encapsulate the task name and completion status.
Let's create a method to get the name of the Task as a String.

This will help us add other properties of Tasks in future versions (like deadlines)

There is no data structure to store Tasks.

Tasks need to be stored and list needs to be displayed.

Let's create a ToDoList class, ToDoList.java, that contains a HashMap(with variable name toDoList) with task index as the key and the Task as the value.
Let's create a method to add a Task to the list when a String is provided.
Let's create a method to display the index and name of all Tasks in the Hashmap.

This allows us to add other functionalities of the ToDoList in future versions (like deletion of tasks)

SirDuke class prints all commands received.

SirDuke class receives, executes and implements the execution of each command, and this complicates
the function of the SirDuke class.

Let's create a switch statement to call the appropriate method based on the given command.
Let's remove unnceccessary print statements.

We create a switch statement to make it easy to add code for other commands to be handled by the SirDuke class in future versions.
We move implementation of execution of commands to the ToDoList class so that SirDuke class only has to receive command
and decide the corresponding action to take, this categorises the functions of each class and makes it easier to know where to
write code when writing features in future versions.

HORIZONTAL_LINE is common across multiple classes.

We have to create a HORIZONTAL_LINE every time we intend to use it in a class.

Let's create a Format class, Format.java that contains all objects we intend to use for formatting printed output.

This allows us to add other formatting objects and use them across all classes that need them without having to re-define them
every time we need to use them.

noahang committed:
Revert "Add Task.java, ToDoList.java, Format.java"

This reverts commit 527cb777b4491bf3e59eb245e120e8dd67d03ecc.

noahang committed:
Reapply "Add Task.java, ToDoList.java, Format.java"

This reverts commit 8917de17cbda06b520f04d129f2b085245b679db.

noahang committed:
Remove test files, stop tracking test files

noahang committed:
edit Task.java class

add javadoc for for Task class
taskName member changed to string description
status member changed to boolean isDone
both members made protected
getTaskName() method edited accordingly

noahang committed:
add javadoc for Task.java, edit string representation of Task.java class using toString, change toDoList.java to be compatible with string representation of Task.java

noahang committed:
add javadoc, add markTaskAsDone() method

noahang committed:
add javadoc for markTaskAsDone() method

noahang committed:
add mark, unmark features to SirDuke.java and complete implementations of mark and unmark in Task.java and ToDoList.java, add error handling for invalid mark and unmark commands

noahang committed:
add ToDoTask.java, add DeadlineTask.java, add EventTask.java, move all of them and Task.java into backend.Task package, import backend.Task package to ToDoList.java

noahang committed:
add command parsing for todo, deadline and event to SirDuke.java, add implementation of adding todo, deadline and event to list to ToDoList, add user guide for chatbot to SirDuke.java, some style fixes to EventTask.java  and ToDoList.java

noahang committed:
add error handling for incomplete add task commands, add error handling for invalid commands

noahang committed:
add style fixes for ToDoTask.java, DeadlineTask.java, EventTask.java, SirDuke.java, add removeTask() method in ToDoList.java, make Task.java an abstract class

noahang committed:
replace toDoList which is a HashMap data structure, into an ArrayList data structure, standardise indexing to start from 0 in ToDoList.java toDoList, add remove function in SirDuke.java and error handling for invalid inputs

ToDoList.java:
- The previous commit had already replaced the HashMap with an ArrayList, we now ensure that the user can refer to a task by it's index starting from 1, and the SirDuke.java class will translate it to the index to start from 0, so that the ToDoList class can operate on the indexes of items as starting from 0
- ArrayList instead of HashMap is preferred, so that when a Task is removed from the ToDoList, the Tasks below it will automatically move up an index in the underlying ArrayList data structure and hence in the ToDoList, this moving up of indexes would be alot harder to implement if we continue to use a HashMap

noahang committed:
rename removeTask() method in ToDoList.java to deleteTask(), rename "remove" command as "delete"

noahang committed:
add toString() method for ToDoList.java, add file save feature to SirDuke.java, update SirDuke.start() method to take File listFile as an argument, documentation, update .gitignore to ignore data file sirDuke.txt

noahang committed:
Merge branch 'branch-level-7'

noahang committed:
add IllegalStartAndEndDateException.java

noahang committed:
change String toBeCompletedBy in DeadlineTask.java, String startTime and String endTime in EventTask.java  to variables of same name of type LocalDate to allow for dates to be intepreted, add error handling in EventTask.java and SirDuke.java when startTime is after endTime, minor style fixes

noahang committed:
Merge branch 'branch-Level-8'

noahang committed:
add runtest.bat, add EXPECTED.TXT, add input.txt automated testing for todo command, list command and bye command, minor bug fix in ToDoList.java by changing ArrayList() to ArrayList<>(), remove FORMAT.HORIZONTAL_LINE import statement and change it to a constant within each class instead.

please take note to change filepathname to be the cwd "./data/sirDuke.txt" when testing within INTELLIJ. To run the testing script, we specify the file pathname as "../data/sirDuke.txt"

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+
+/**
+ * Class that acts as the User Interface.
+ *<p>
```
```suggestion
 * <p>
```

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+ *
+ *      - To close the chatbot, use the following command:
+ *      bye
+ *</p>
```
```suggestion
 * </p>
```

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+     * Executes commands accordingly.
+     * Updates and saves to listFile only if a command is SUCCESSFULLY EXECUTED.
+     *
+     * @param listFile the file for the contents of the ToDoList to be written into
```
The sentences should start with capitalisation, here I am referring to https://nus-cs2103-ay2425s2.github.io/website/schedule/week3/topics.html#implementation-documentation-tools-javadoc-what as a reference.
This is found in the other files as well.

```suggestion
     * @param listFile The file for the contents of the ToDoList to be written into.
```

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+     * @param listFile the file for the contents of the ToDoList to be written into
+     */
+    public static void start(File listFile) {
+
```
I am not too sure if this is considered a coding standard violation, as line breaks are usually meant to separate code segments based on logic. I think the variables should be put together? 
These line breaks are found in the other files as well.
```suggestion
```

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+    public static void start(File listFile) {
+
+        ToDoList list = new ToDoList();
+
```
```suggestion
```

wj200 replied:
Line spacing could be removed

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+            String regex = "/";
+            String[] parsedCommand = command.split(regex);
+            switch (parsedCommand[0]) { //first word of the command
+                case "bye":
```
Refer to https://se-education.org/guides/conventions/java/intermediate.html for the coding standards.
Switch case should not have indentations.

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+                    try {
+                        int index = Integer.parseInt(parsedCommand[1]) - 1;
+                        list.markTaskAsDone(index);
+                    } catch (ArrayIndexOutOfBoundsException e) { //no index provided after "mark"
```
Comments should have spaces after. `//` 
This is present in the other comments as well.

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+
+        Scanner scanner = new Scanner(System.in);
+
+        while(true) {
```
I am unsure if this will be considered a coding standard violation, but perhaps it would be better to either assign it to a variable or put the scanner inside?

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+            }
+        }
+    }
+    public static void main(String[] args) {
```
Missing line break as well as header comment.

ChinZJ commented on src/main/java/backend/Task/DeadlineTask.java:
```diff
@@ -0,0 +1,33 @@
+import java.time.format.DateTimeParseException;
+
+/**
+ * Represents a task with a deadline. A <code>DeadlineTask<code> object has a
```
I think the `<code>` at the end should be `</code>`.
This was found in the other code files as well. 

```suggestion
 * Represents a task with a deadline. A <code>DeadlineTask</code> object has a
```

ChinZJ commented on src/main/java/SirDuke.java:
```diff
@@ -0,0 +1,210 @@
+import backend.ToDoList;
+
+/**
+ * Class that acts as the User Interface.
```
Good that the comment ends with a full stop.
This should be applied to all other comments.

ChinZJ commented on src/main/java/backend/Task/IllegalStartAndEndDateException.java:
```diff
@@ -0,0 +1,23 @@
+    IllegalStartAndEndDateException(LocalDate startTime, LocalDate endTime) {
+        this.startTime = startTime;
+        this.endTime = endTime;
+    }
```
Missing line break.

ChinZJ commented on src/main/java/backend/ToDoList.java:
```diff
@@ -0,0 +1,143 @@
+        }
+
+    /**
+     * Delete a task in the list.
```
Should be in imperative
```suggestion
     * Deletes a task in the list.
```

ChinZJ commented on src/main/java/frontend/Format.java:
```diff
@@ -0,0 +1,6 @@
+package frontend;
+
```
Header comment missing.

wj200 replied:
Header comment for package missing

ChinZJ reviewed (commented):
LGTM, minor mistakes that needs to be corrected.

noahang committed:
add storage.java

noahang committed:
add methods to ToDoList.java and Task.java to be compatible with Storage.java

noahang committed:
rename ToDoList toDoList to ToDoList taskList, add storage variable

noahang committed:
add constructor for Storage.java

noahang committed:
add constructor for SirDuke.java that initialises Storage storage and ToDoList taskList

noahang committed:
turn SirDuke.java into a non-static class, remove duplicate save file code

noahang committed:
add javadoc for Storage.java

noahang committed:
rename ToDoList list to ToDoList taskList

noahang committed:
repackage IllegalStartAndEndDateException into exceptions, rename backend.Task package to backend.task

noahang committed:
make IllegalStartAndEndDateException public

noahang committed:
replace toString method with getStatusIcon and toFileEntry method

this allows us to write tasks as file entries into the hard disk in a standardised format

noahang committed:
add implementation for toFileEntry() method in EventTask.java

noahang committed:
add implementation for toFileEntry() method in DeadlineTask.java, ToDoTask.java

wj200 commented:
Overall, LGTM

as the other PR reviewer has commented. Just some minor coding standard violations

noahang committed:
extract functionality of command parsing into Parser.java class, extract functionality of UI from all other classes into SirDuke.java, add Main class

noahang committed:
refactor Main.java, extract methods from ToDoList.java to SirDuke.java

noahang committed:
Merge branch 'add-gradle-support' of https://github.com/noahang/ip

noahang committed:
add build and run using gradle

noahang committed:
fix major bug where Parser only reads one command and exits

noahang committed:
add junit tests, create jar file

noahang committed:
add javaFX dependencies to build.gradle

noahang committed:
bug fix: read list data from disk

noahang committed:
bug fix: read and write from list, bug fix: command reading

noahang committed:
style fixes

noahang committed:
add documentation for Task.toFileEntry() method

noahang committed:
Merge branch 'A-CodingStandard'

noahang committed:
update documentation for deleteTask, unmarkTaskAsDone, markTaskAsDone

noahang committed:
add find command

noahang committed:
Merge branch 'Level-9'

noahang committed:
refactor every class into a package, add javaFX GUI that echoes user input

noahang committed:
add UI.java class

noahang committed:
add JavaFX gui, extract SirDuke methods into Commands

noahang committed:
add javaFX gui improvements

noahang committed:
Merge branch 'Level-10'

noahang committed:
remove unncessary import statements

noahang committed:
add assertions for createTask commands

noahang committed:
remove unncecessary import statements, reorder necessary import statements

noahang committed:
reorder exception handling

noahang committed:
reorder exception handling, minor style fixes

noahang committed:
remove unnecessary print statements

noahang committed:
Merge pull request #2 from noahang/branch-A-Assertions

Branch a-assertions

noahang committed:
Merge branch 'master' of https://github.com/noahang/ip

noahang committed:
Merge branch 'master' into branch-A-CodeQuality

noahang committed:
Merge pull request #3 from noahang/branch-A-CodeQuality

Branch a-code quality, improve code quality

noahang committed:
Merge branch 'master' of https://github.com/noahang/ip

noahang committed:
add update function

noahang committed:
extract EditCommand's execute() cases into  4 different classes

EditCommand's execute() method parses the type of edit to give editType (which should be done by the parser) and executes based on the editType, this results in a deeply nested code and a very long method.

Refactoring this command into 4 different command classes allows the parser to determine which type of edit command to generate, while shortening the execute() method.

Let's:
* Remove the EditCommand class
* Add 4 different classes EditDescriptionCommand for ("/description"), EditToBeCompletedByCommand for ("/by"), EditStartTimeCommand for ("/start)", EditEndTimeCommand for ("/end)", each with an execute() method containing the functions of the respective of the 4 cases of editType
* Edit the Parser's parse() method to create each of these commands given their respective cases
* Allow ToDoList's getTask() method to throw an IndexOutOfBoundsException so that we know if we failed to get a task and allow each of our new command Classes to handle the error by informing that the task does not exist

noahang committed:
Allow SirDuke to add time to event and deadline tasks

The startTime, endTime and toBeCompletedBy fields in EventTask and DeadlineTask are of type LocalDate and can only read dates but not times.

Changing these fields to LocalDateTime allows user to specify time, making the task descriptions more useful. We will henceforth call fields containing both dates and times as time fields.

Let's:
* Replace the types of the startTime, endTime and toBeCompletedBy fields from LocalDate to LocalDateTime.
* Create a parsingFormatter field in the Task class to describe the format to be used for String inputs to be parsed into LocalDateTime.
* Create a printingFormatter field in the Task class to describe the format to be used for when each of the time fields are printed to String
* Edit the toFileEntry() methods of EventTask and DeadlineTask to follow the parsing format described by parsingFormatter when converting times of the tasks to file entries, so that it can be readDataFromDisk by the Storage class.
* Edit the EditEndTimeCommand and EditStartTime command, setEndTime() and setStartTime() for the EventTask class to work with LocalDateTime, while also preventing user from setting startTime or endTime to be invalid by throwing an IllegalStartAndEndTimeException.
* Edit the IllegalStartAndEndTimeException to print dates in the same format as that described in printingFormatter.

noahang committed:
Add javadoc and header comments for all non-trivial classes and methods

noahang committed:
Add UI.png

noahang committed:
Add Junit tests for DeadlineCommand and ToDoList

Add test for execute() method of DeadlineCommand and createToDoTask() method of ToDoList. Refactor all classes into seedu package to allow for JUnit tests to work.

noahang committed:
Remove unnecessary comment from EditToBeCompletedByCommand.java

noahang committed:
Refactor build.gradle's mainClass into seedU so gradle can build and run

noahang committed:
Improve find, delete and bye GUI responses

noahang committed:
Improve find, delete and bye GUI responses

noahang committed:
Add Userguide

noahang committed:
Minor style fix

noahang committed:
Merge remote-tracking branch 'origin/master'

noahang committed:
Rename Ui.png
